Adaptive patch2self

captainnova opened the pull request:
Hello, this will probably be most interesting to @ShreyasFadnavis .

Shreyas, I think I figured out why p2s is blurring in diffusion space, and mostly fixed it. The problem is that by training the regressor to predict an entire volume it is assuming that all of the voxels have the same "diffusion response function", i.e. it is like saying they have the same diffusion tensor. The result looks better than that, though, because the regressor is free to give more weight to similar volumes, so even though it is using a one-size-fits-all interpolation kernel, it does not have to interpolate from very far away.

My solution is to train different regressors for different types of voxels, and then for any given voxel use a weighted sum of the regressors. The "types of voxels" comes from either Independent Component Analysis or Principal Component Analysis (ICA is slightly better), and the voxel's position in component space also sets the weights for the regressors. Thus each voxel gets a regressor that is appropriate to it, preserving signal without hurting denoising by going to the extreme of giving each voxel a regressor that is exactly tuned to it.

Since the equivalent of a "patch" is now a collection of voxels with similar intensities for each volume, regardless of their spatial location, I took out patch_radius. I now think p2s really is better working only in diffusion space, and the improvement I was seeing before with larger patches was just from the regressor getting more variables to work with. Because the interface changed, I ended up making adaptive_patch2self its own command instead of trying to insert this into p2s.

This will probably make more sense with pictures, but I'll submit the pull request for review now and add images later instead of trying to do everything at once.

Best wishes,

Rob

pep8speaks commented:
Hello @captainnova, Thank you for submitting the Pull Request !

* In the file [`dipy/denoise/adaptive_patch2self.py`](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py):

> [Line 28:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L28): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 80 characters)
> [Line 51:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L51): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)
> [Line 69:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L69): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 80 characters)
> [Line 71:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L71): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 80 characters)
> [Line 76:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L76): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 80 characters)
> [Line 80:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L80): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 80 characters)
> [Line 84:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L84): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)
> [Line 89:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L89): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 80 characters)
> [Line 116:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L116): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 80 characters)
> [Line 120:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L120): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 80 characters)
> [Line 127:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L127): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 80 characters)
> [Line 139:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L139): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 80 characters)
> [Line 152:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L152): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 80 characters)
> [Line 153:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L153): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 80 characters)
> [Line 196:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L196): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 80 characters)
> [Line 197:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L197): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 80 characters)
> [Line 199:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L199): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 80 characters)
> [Line 200:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L200): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 80 characters)
> [Line 226:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L226): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 80 characters)
> [Line 228:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L228): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 80 characters)
> [Line 229:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L229): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 80 characters)
> [Line 233:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L233): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 80 characters)
> [Line 242:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L242): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 80 characters)
> [Line 277:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L277): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 80 characters)
> [Line 285:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L285): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (97 > 80 characters)
> [Line 291:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L291): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 80 characters)
> [Line 361:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L361): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 80 characters)
> [Line 429:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L429): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 80 characters)
> [Line 438:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L438): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 80 characters)
> [Line 446:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L446): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 80 characters)
> [Line 447:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L447): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 80 characters)
> [Line 502:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L502): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 80 characters)
> [Line 503:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L503): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 80 characters)
> [Line 508:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L508): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 80 characters)
> [Line 509:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L509): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 80 characters)
> [Line 517:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L517): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 80 characters)
> [Line 522:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L522): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 80 characters)
> [Line 523:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L523): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 80 characters)
> [Line 524:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L524): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 80 characters)
> [Line 563:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L563): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 80 characters)
> [Line 610:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L610): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 80 characters)
> [Line 672:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/adaptive_patch2self.py#L672): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 80 characters)

* In the file [`dipy/denoise/patch2self.py`](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/patch2self.py):

> [Line 171:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/patch2self.py#L171): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 80 characters)
> [Line 351:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/patch2self.py#L351): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 80 characters)

* In the file [`dipy/denoise/tests/test_adaptive_patch2self.py`](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/tests/test_adaptive_patch2self.py):

> [Line 22:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/tests/test_adaptive_patch2self.py#L22): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 80 characters)
> [Line 27:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/tests/test_adaptive_patch2self.py#L27): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 80 characters)
> [Line 30:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/tests/test_adaptive_patch2self.py#L30): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 80 characters)
> [Line 38:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/tests/test_adaptive_patch2self.py#L38): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 80 characters)
> [Line 39:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/tests/test_adaptive_patch2self.py#L39): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 80 characters)
> [Line 44:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/tests/test_adaptive_patch2self.py#L44): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 80 characters)
> [Line 47:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/tests/test_adaptive_patch2self.py#L47): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 80 characters)
> [Line 50:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/tests/test_adaptive_patch2self.py#L50): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 80 characters)
> [Line 57:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/tests/test_adaptive_patch2self.py#L57): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 80 characters)
> [Line 59:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/denoise/tests/test_adaptive_patch2self.py#L59): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 80 characters)



* In the file [`dipy/workflows/denoise.py`](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/workflows/denoise.py):

> [Line 91:81](https://github.com/dipy/dipy/blob/a6599339d7c7813d4fe4b85cad7b518a0a125ba7/dipy/workflows/denoise.py#L91): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 80 characters)

Do see the [DIPY coding Style guideline](https://github.com/dipy/dipy/blob/master/doc/devel/coding_style_guideline.rst)

captainnova commented:
I forgot to mention that another reason to drop patch_radius is that I added masking, which gives a big speed boost. So altogether with a few x speedup from masking, and a couple of dozen more regressors, adaptive_patch2self is only several x slower than patch2self.

ShreyasFadnavis commented:
@captainnova:
Hi Rob, this is very good work and thanks a lot for the PR! I am looking at your code and looks quite good to me. I like the idea name Patch2Self Adaptive too 👍🏽 

I only briefly looked at the code and realized that it could be integrated withing Patch2Self as an option. We can of course keep the workflow separate -- which looks good to me. For the `patch_radius=0` I was going to downgrade that in fashion to what you have too! 

I also have some new work on speeding up Patch2Self which would work with your modification too :) I will try the code out ASAP (especially with the data that you have). Can you also add a small example of how I can run the code and reproduce your results... So that we are on the same page? It will help with reviewing the PR too 👍🏽 

As a side question: do you get the same result with and without masking? If yes -- That is quite neat!

captainnova commented:
@ShreyasFadnavis :

Hi Shreyas, thanks for the kind words.

> I only briefly looked at the code and realized that it could be easily integrated withing Patch2Self as an option.

That would make sense to me too, and was my original intent, but when the urge to get rid of patch_radius grew too strong I ended up splitting it into a new command, because I didn't want to "break" p2s. As the original author of p2s, though, you have more freedom in choosing what changes to introduce.

> The one that is in SKlearn is slowing things down I assume.

I haven't compared different implementations of the regressions or decompositions for speed. Using DGESVD for PCA and sklearn for ICA is a little weird, since sklearn also comes with PCA, and has a simpler/more consistent interface. But I started with PCA, and copied the example in denoise_lpca. When I added ICA I wondered about switching PCA to sklearn, but for all I know dipy has a reason to use DGESVD instead.

> I also have some new work on speeding up Patch2Self which would work with your modification too :) I will try the code out ASAP (especially with the data that you have). Can you also add a small example of how I can run the code and reproduce your results... So that we are on the same page? It will help with reviewing the PR too 👍🏽

Running it is really easy, and very similar to p2s. I added a dipy_denoise_adaptive_patch2self CLI (which I'm not sure I've used, to be honest), but here is the python gist:
```python
import dipy.denoise.adaptive_patch2self as ap2s
from dipy.io.image import load_nifti, save_nifti
import numpy as np

data, affine = load_nifti('raw.nii')
bvals = np.loadtxt('raw.bval')

# I use https://github.com/captainnova/dmri_segmenter to make these
mask, _ = load_nifti('raw_TIV.nii')

dnarr = ap2s.adaptive_patch2self, data, bvals, mask=mask, out_dtype=np.float32, verbose=True)
save_nifti('dnlinica.nii', dnarr, affine)
```

Properly evaluating the results takes more code, though. It is especially important to look at how the residuals change w.r.t. volume, not just the spatial pattern. Unfortunately the unwanted correlations do not show up very well in either movies of the residuals or time series of individual voxels. They do show up very well, and can be quantified, in maps of Pearson's r vs. signal, but that takes some code. What is the recommended way to share code that is used to investigate how a feature works, but is not needed by the feature itself?

> As a side question: do you get the same result with and without masking? If yes -- That is quite neat!

Almost - I think the result is slightly better with masking. What happens without masking is the first component is "is it air?", which is not very interesting for denoising. With masking that component goes away, and the others get promoted by one. But with 8-12 component axes being used in total this does not make a huge difference. It would probably make a bigger difference for original p2s. Without masking it is mostly learning the difference between air and not air, but with masking it would effectively be one component ahead of where it is now. You could simulate this by using ap2s(with and without masking, n_comps=0).

skoudoro reviewed (commented):
Hi @captainnova,

Thank you for this new feature! 

I would like to know what is the next step concerning this PR @ShreyasFadnavis and @captainnova.

+1 for integrating within Patch2Self as an option

Could you precise if this PR is still in work in progress or not? 

If not, Could you add a short tutorial? It would make it easier to test. 

It would be good if you could improve the consistency of your docstring in general. Sometimes, it is really good, and sometimes inexistent. See below for a really quick review (just code style) until I know more about the status.

Thank you!

skoudoro commented:
Also, Do not forget to rebase your PR since there is currently some conflict with the codebase.

Thank you !

captainnova commented:
Hi @skoudoro,

> I would like to know what is the next step concerning this PR @ShreyasFadnavis and @captainnova.
> 
> +1 for integrating within Patch2Self as an option

That's exactly what @ShreyasFadnavis is working on. My PR was mostly a feature addition, but because I also took a couple of things out, I implemented it as a separate command. Shreyas agreed that patch2self no longer needs the removed items either, so adaptive_patch2self and patch2self are being merged.

I agree with your code style comments, but think it is best to wait for Shreyas to complete his merge + incorporate your comments + rebase, and then see what remains to be done.

captainnova commented on dipy/denoise/adaptive_patch2self.py:
```diff
@@ -0,0 +1,699 @@
+    warn(w)
+
+
+def site_weight_beam(U, pca_ind, sign, sigma, growth_func):
+    """Return weights that increase going along the sign direction of
+    U[:, pca_ind], and fall off Gaussianly going away from that axis.
+    """
```
agreed

captainnova commented on dipy/denoise/adaptive_patch2self.py:
```diff
@@ -0,0 +1,699 @@
+    return site_weight_beam(U, pca_ind, sign, sigma, np.arctan)
+
+
+def getRandomState(seed):
```
I have no objection to underscored_function_names. @ShreyasFadnavis ?

captainnova commented on dipy/denoise/adaptive_patch2self.py:
```diff
@@ -0,0 +1,699 @@
+        The principal components in feature space as orthonormal row vectors,
+        matching the order of S.
+    """
+    U, S, Vt = svd(X, *svd_args)[:3]
```
I agree, but I copied this from localpca.py thinking it was dipy's preferred way of doing PCA/SVD. If there's no performance benefit to doing it that way, using sklearn.decomposition would provide a better interface.

skoudoro replied:
So, good catch! 😄  we need to change this in `localpca.py` also

captainnova commented on dipy/denoise/adaptive_patch2self.py:
```diff
@@ -0,0 +1,699 @@
+    ----------
+    X : (nsamples, nfeatures) array
+        The array to decompose.
+    n_comps : int or None
```
OK

captainnova commented on dipy/denoise/adaptive_patch2self.py:
```diff
@@ -0,0 +1,699 @@
+    return ica.fit(X).transform(X)
+
+
+class AdaptivePatch2Self(object):
```
OK, I'm still necessarily in the habit of 2/3 compatibility. That's a nice link, gathering a lot of info in one place.

captainnova commented on dipy/denoise/adaptive_patch2self.py:
```diff
@@ -0,0 +1,699 @@
+    fit the voxel, which would prevent denoising.
+    """
+    def __init__(self, data, n_comps=None, dtype=np.float32, model='ols',
+                 mod_kwargs={}, site_weight_func=site_weight_beam_linear,
```
Good to know!

skoudoro commented:
ok, great! Thank you for the information! So let's wait for Shreyas, I will tag this PR as WIP to avoid any review before this update.

ShreyasFadnavis commented:
> Hi @skoudoro,
> 
> > I would like to know what is the next step concerning this PR @ShreyasFadnavis and @captainnova.
> > +1 for integrating within Patch2Self as an option
> 
> That's exactly what @ShreyasFadnavis is working on. My PR was mostly a feature addition, but because I also took a couple of things out, I implemented it as a separate command. Shreyas agreed that patch2self no longer needs the removed items either, so adaptive_patch2self and patch2self are being merged.
> 
> I agree with your code style comments, but think it is best to wait for Shreyas to complete his merge + incorporate your comments + rebase, and then see what remains to be done.

+1 to what Rob said! I am testing the code and also working on integrating it with the current implementation. 
@skoudoro thank you for the review 💯 I will make my changes on top of this branch as soon as I have it working 👍🏽

skoudoro commented:
Hi @captainnova, Hi @ShreyasFadnavis,

We plan to talk about your PR during the DIPY online meeting tomorrow (October 7th, 2021 | 1pm EST / 7pm CET / 10am PT).

It would be great if you could attend this meeting to provide us a short update. 

More information on https://github.com/dipy/dipy/discussions/2398

Thank you!
